### ISSUE TITLE
[FEATURE] add more detail to stateful testing docs

### ISSUE THREAD

**Issue by mdavis-xyz** [schemathesis/schemathesis#803]
**Is your feature request related to a problem? Please describe.**

Looking at the documentation for stateful testing, it's not clear whether schemathesis can do exactly what I want it to do.

**Describe the solution you'd like**

For my API, I want the client to create a resource with one call. In my spec there is a link there to a second call, to check the status of the resource.

Schemathesis can automatically take the resource ID returned by the first call, and put it in the path for the second call.
That's great! Super useful.

But how to I assert things about the second call? e.g. assert that the second call has a particular value in the response, of that the HTTP status of the second call is one particular status out of the two defined in the spec?

## Example

I want to test POST to `/requests`.
The response will either be status 200 OK or 400 (not ok).
In both cases I want schemathesis to then do a `GET` to `/requests/{requestId}` to check the status.
Then I want to assert that:

```
assert ((calls[0].http_status == 200) and (calls[1].body in ['IN_PROGRESS', 'COMPLETE'])) or )(calls[0].http_status == 400) and (calls[1].body == 'REJECTED'))
```

```
  /requests:
    post:
      summary: Submit a request
      operationId: createRequest
      requestBody:
        required: true
        description: User object
        content:
          application/json:
            schema:
              $ref: '#/components/schemas/User'
      responses:
        '200':
          summary: Request accepted
          ...
          links:
            GetUserByUserId:
              operationId: getRequestStatus
              parameters:
                requestId: '$response.body#/id'
        '400':
          summary: Request rejected
          ...
          links:
            GetUserByUserId:
              operationId: getRequestStatus
              parameters:
                requestId: '$response.body#/id'
  /requests/{requestId}:
    get:
      summary: Gets the status of a request by ID
      operationId: getRequestStatus
      parameters:
        - in: path
          name: requestId
          required: true
          schema:
             type: string
      responses:
        '200':
          content:
            application/json:
              schema:
                type: string
                  enum:
                  # I want to assert that 
                  # if the first call returned HTTP status 200,
                  # the second call should return one of these:
                  - IN_PROGRESS
                  - COMPLETE
                  
                  # I want to assert that
                  # If the first call returned HTTP status 400,
                  # the second call should return this
                  - REJECTED
```

Is this possible with schemathesis today?

**Comment by Stranger6667** [comment]
Hi @mdavis-xyz ,

> But how to I assert things about the second call?

I think that now, Schemathesis doesn't provide a direct way to make checks that depend on the result of the previous call. 

> Is this possible with schemathesis today?

It is possible by using the underlying hypothesis strategies instead. Unfortunately, in this case, you'll need to extract the data manually. It might look something like this:

```python
from hypothesis import given

schema = schemathesis.from_uri("http://example.com/openapi.json")

create_request = schema["/requests"]["POST"].as_strategy()
get_request = schema["/requests/{requestId}"]["GET"].as_strategy()

@given(
  create=schema["/requests"]["POST"].as_strategy(),
  get=schema["/requests/{requestId}"]["GET"].as_strategy(),
)
def test_api(create, get):
    create_response = create.call()
    create.validate_response(create_response)
    get.path_parameters["requestId"] = create_response.json()["id"]
    get_response = get.call()
    get.validate_response(get_response)
    if create_response.status_code == 200:
        assert get_response.json() in ("IN_PROGRESS", "COMPLETE")
    if create_response.status_code == 400:
        assert get_response.json() == "REJECTED"
```

The example above is quite wordy and is specific only to using Schemathesis in Python tests - there is no direct way to do the same with the CLI.

However, I think that the example above might be much simpler if we'd implement some `source_response` attribute that will point to the original response:

```python
...

@schema.parametrize(stateful=Stateful.links)
def test_api(case):
    response = case.call()
    case.validate_response(response)
    if case.path == "/requests/{requestId}" and case.source_response is not None:
        if case.source_response.status_code == 200:
            assert response.json() in ("IN_PROGRESS", "COMPLETE")
        if case.source_response.status_code == 400:
            assert response.json() == "REJECTED"
```

And then for the CLI use case, it might be achieved with a custom check:

```python
@schemathesis.register_check
def conditional_response_value(response, case):
    if case.path == "/requests/{requestId}" and case.source_response is not None:
        if case.source_response.status_code == 200:
            assert response.json() in ("IN_PROGRESS", "COMPLETE")
        if case.source_response.status_code == 400:
            assert response.json() == "REJECTED"
```

And then `schemathesis --pre-run=path.to.check_module run -c conditional_response_value ...` will do the trick. 

I hope that the first example may somehow be useful for your situation. And that possible "source" attribute seems like a no big deal to implement - I'll take a look.

Besides this, I'll update the docs - but probably as a part of the implementation of this attribute.

Let me know if it works for you :) If now, we might think about some alternative

**Comment by Stranger6667** [comment]
Resolved via #815.

Check the new documentation - https://schemathesis.readthedocs.io/en/stable/stateful.html

The use case you described is explicitly supported now :)
